### PR TITLE
expression: fix cast negint to string

### LINKED ISSUE #44786: CastIntAsString did not work in some cases
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `t` (
  `id` int(11) DEFAULT NULL,
   `var` int(1) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into t values(1,-1);
select id, var, cast(var as char) from t;
```

### 2. What did you expect to see? (Required)

```
id | var | cast(var as char)
-- | -- | --
1 | -1 | -1
```

### 3. What did you see instead (Required)

```
id | var | cast(var as char)
-- | -- | --
1 | -1 | -
```

### 4. What is your TiDB version? (Required)

v6.5.2

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: close #44786 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -291,8 +291,13 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	argTp := args[0].GetType().EvalType()
 	switch argTp {
 	case types.ETInt:
+		// check https://github.com/pingcap/tidb/issues/44786
 		if bf.tp.GetFlen() == types.UnspecifiedLength {
-			bf.tp.SetFlen(args[0].GetType().GetFlen())
+			if args[0].GetType().GetType() == mysql.TypeLong {
+				bf.tp.SetFlen(11)
```
Comment (tiancaiamao) [inline expression/builtin_cast.go:299]: It's better to add some comment here.
This change looks like magic ...

**Review by tiancaiamao — APPROVED** [review]

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -292,7 +292,14 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
...
 	case types.ETInt:
 		if bf.tp.GetFlen() == types.UnspecifiedLength {
-			bf.tp.SetFlen(args[0].GetType().GetFlen())
+			// check https://github.com/pingcap/tidb/issues/44786
+			// set flen from integers may truncate integers, e.g. char(1) can not display -1[int(1)]
+			// set it to 11 as mysql
+			if args[0].GetType().GetType() == mysql.TypeLong {
+				bf.tp.SetFlen(11)
+			} else {
+				bf.tp.SetFlen(args[0].GetType().GetFlen())
```
Comment (XuHuaiyu) [inline expression/builtin_cast.go:301]: Why not set it to `GetFlen()+1` ?

Comment (xhebox) [reply expression/builtin_cast.go:301]: `char(2)` can not display `-127 | int(1)`.

**Review by wshwsh12 — CHANGES_REQUESTED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**